sdk/servicebus/azure-servicebus/azure/servicebus/aio/_async_auto_lock_renewer.py

```diff
@@ -0,0 +1,130 @@
+        self._shutdown = asyncio.Event()
+        self._futures = []
+        self.loop = loop or get_running_loop()
+        self.sleep_time = 1
+        self.renew_period = 10
+
+    async def __aenter__(self) -> "AutoLockRenew":
+        if self._shutdown.is_set():
+            raise ServiceBusError("The AutoLockRenew has already been shutdown. Please create a new instance for"
+                                  " auto lock renewing.")
+        return self
+
+    async def __aexit__(self, *args: Iterable[Any]) -> None:
+        await self.close()
+
+    def _renewable(self, renewable: "Union[ReceivedMessage, ServiceBusSession]") -> bool:
+        if self._shutdown.is_set():
+            return False
+        if hasattr(renewable, '_settled') and renewable._settled:
+            return False
+        if renewable._lock_expired:
+            return False
+        if not renewable._receiver._running:
+            return False
+        return True
+
+    async def _auto_lock_renew(self,
+                               renewable: "Union[ReceivedMessage, ServiceBusSession]",
+                               starttime: datetime.datetime,
+                               timeout: int,
+                               on_lock_renew_failure: "Optional[Callable[[Union[ServiceBusSession, ReceivedMessage], Optional[Exception]], Awaitable[None]]]"=None) -> None:
+        _log.debug("Running async lock auto-renew for %r seconds", timeout)
+        error = None
+        clean_shutdown = False # Only trigger the on_lock_renew_failure if halting was not expected (shutdown, etc)
+        try:
+            while self._renewable(renewable):
+                if (utc_now() - starttime) >= datetime.timedelta(seconds=timeout):
+                    _log.debug("Reached auto lock renew timeout - letting lock expire.")
+                    raise AutoLockRenewTimeout("Auto-renew period ({} seconds) elapsed.".format(timeout))
+                if (renewable.locked_until_utc - utc_now()) <= datetime.timedelta(seconds=self.renew_period):
+                    _log.debug("%r seconds or less until lock expires - auto renewing.", self.renew_period)
+                    await renewable.renew_lock()
+                await asyncio.sleep(self.sleep_time)
+            clean_shutdown = not renewable._lock_expired
+        except AutoLockRenewTimeout as e:
+            renewable.auto_renew_error = e
+            clean_shutdown = not renewable._lock_expired
+        except Exception as e:  # pylint: disable=broad-except
+            _log.debug("Failed to auto-renew lock: %r. Closing thread.", e)
+            error = AutoLockRenewFailed(
+                "Failed to auto-renew lock",
+                inner_exception=e)
+            renewable.auto_renew_error = error
+        finally:
+            if on_lock_renew_failure and not clean_shutdown:
+                await on_lock_renew_failure(renewable, error)
+
+    def register(self,
```

Comment:
I would ask a dumb question, but this is always confusing me when I'm in the async world:

Should we make this `async def register(self,...)` to keep consistent even though the implementation doesn't call any coroutine method?

I think in async world, users tend to use `await`, so some of them might take it as granted put `await` in front of `register` and what they finally got would be `object NoneType can't be used in 'await' expression`.

Reply:
this is a non-blocking question for this PR as in track 1 it's already there.

Reply:
While I tend to agree with you in terms of the consistency factor, this would beg much larger changes across our lib (e.g. get_queue_receiver is not awaitable, from_connection_string, etc) which I'm glad to make but would want to check beforehand.

This isn't defined in our guidelines, and looking at precedent in storage/otherwise didn't turn up any obvious clues; @annatisch seeking your wisdom; see Adam's question above.

Reply:
Synced with Anna: We don't actually have a principle to force this to be async.  Will be keeping this as is.  (Don't take this as a critique of your thinking, as said I tend to agree, but we should align here so I'll be keeping this as is)